src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1308,6 +1308,7 @@ private BoundExpression GetDefaultParameterSpecialNoConversion(SyntaxNode syntax
             // Params array is filled in the local rewriter
             var lastIndex = expanded ? ^1 : ^0;
 
+            var arguments = argumentsBuilder.ToImmutable();
```

Comment:
>var arguments = argumentsBuilder.ToImmutable(); [](start = 12, length = 47)

Do we really need this allocation? It feels like we could pass the builder to the local function.

Reply:
Yes I think we can do that.

Reply:
> Do we really need this allocation? It feels like we could pass the builder to the local function.

This changes behavior for cases similar to the following

```csharp
using System;
using System.Runtime.CompilerServices;

C.M();

public static class C
{
    public static void M(
        [CallerMemberName] string callerName = "",
        [CallerArgumentExpression(""callerName"")] string argumentExp = "")
    {
        Console.WriteLine(callerName);

        // If we pass the builder, this will eventually get the value that was auto-generated.
        // However, if converted to immutable array at the beginning, we'll be sure that only arguments that are explicitly passed are used. This test case is already a PROTOTYPE as I'm not sure which behavior you'll want.
        // So basically, do we want `argumentExp` to have the value of `C.M()` or not?
        Console.WriteLine(argumentExp);
    }
}
```

Reply:
> If we pass the builder, this will eventually get the value that was auto-generated.

I believe there are other ways to guard against that without allocating an extra array. For example by simply remembering how many explicit arguments we have.

---
In reply to: [600279189](https://github.com/dotnet/roslyn/pull/51952#discussion_r600279189) [](ancestors = 600279189)

Reply:
Done